### PR TITLE
feat: add noname as a frontend to sonobe

### PR DESCRIPTION
**Pull request by dmpierre**
This PR adds [noname](https://github.com/zksecurity/noname/tree/main) as a frontend to sonobe. It includes:
- The implementation of the `FCircuit` trait for a noname circuit
- Utilities to deal with noname's particular data structures
- A "full" flow example, similar to what has already been done for circom

cc @mimoo

### CODE REVIEW

File: examples/noname_full_flow.rs
Comment (dmpierre) [inline examples/noname_full_flow.rs:1]: This is mostly a copy past from `circom_full_flow.rs`

File: folding-schemes/Cargo.toml [hunk 1]
```diff
@@ -23,9 +23,13 @@ num-integer = "0.1"
 color-eyre = "=0.6.2"
 ark-bn254 = {version="0.4.0"}
 ark-groth16 = { version = "^0.4.0" }
+ark-noname = { git = "https://github.com/dmpierre/ark-noname", branch="feat/sonobe-integration" }
```
Comment (dmpierre) [inline folding-schemes/Cargo.toml:26]: This branch makes it possible to change a noname r1cs into an arkworks one, while taking into account the fact that the `cs` has already been assigned witnesses and public io.

File: folding-schemes/src/frontend/noname/mod.rs [hunk 2]
```diff
@@ -0,0 +1,228 @@
...
+            true => {
+                let json = serde_json::to_string(&NoNameJSONExternalInputs::from_fpvars(
+                    &external_inputs.clone(),
+                )?)
+                .map_err(|_| SynthesisError::Unsatisfiable)?;
+                Some(json)
+            }
+        };
+
+        let ivc_inputs = serde_json::to_string(&NoNameJSONIVCInput::from_fpvars(&z_i)?)
```
Comment (dmpierre) [inline folding-schemes/src/frontend/noname/mod.rs:107]: There is a bunch of transformation that we need to do. The flow is: we receive the state and external inputs as `FpVars`, we turn those into `F` elements, those are converted into properly formatted string json inputs (using `serde_json::to_string`) and are finally turned into a data structure that is used as inputs for `generate_witness`.

File: folding-schemes/Cargo.toml [hunk 3]
```diff
@@ -23,9 +23,13 @@ num-integer = "0.1"
...
+noname = { git = "https://github.com/dmpierre/noname" }
+serde_json = "1.0.85"                                                                # to (de)serialize JSON
 
 # tmp imports for espresso's sumcheck
 espresso_subroutines = {git="https://github.com/EspressoSystems/hyperplonk", package="subroutines"}
+serde = "1.0.203"
```
Comment (arnaucube) [inline folding-schemes/Cargo.toml:32]: does this one fall under the `# tmp imports for espresso's sumcheck` imports section?
since we're at it, if you can also update the old comment from `imports` to `import` (singular), thanks ^^

File: examples/noname_full_flow.rs
Comment (arnaucube) [reply examples/noname_full_flow.rs:1]: Cool! Makes sense, while I'm pro avoiding code repetition by default, I think that for the lib examples is good to have the simplest code for new devs to see how to use it, which implies avoiding abstracting methods away and reusing code.

File: folding-schemes/src/frontend/noname/utils.rs [hunk 4]
```diff
@@ -0,0 +1,157 @@
...
+    let yy = external_inputs[1] * ivc_input[1];
+    assert_eq(yy, xx);
+    return [xx, yy];
+}";
+
+pub const NONAME_CIRCUIT_NO_EXTERNAL_INPUTS: &str =
+    "fn main(pub ivc_input: [Field; 2]) -> [Field; 2] {
+    let out = ivc_input[0] * ivc_input[1];
+    return [out, ivc_input[1]];
+}";
```
Comment (arnaucube) [inline folding-schemes/src/frontend/noname/utils.rs:23]: These circuits are used for testing only but are defined as global constants, since they are not part of the lib itself but for the tests only, maybe they could be moved under the `noname/mod.rs# mod tests{`?

File: folding-schemes/src/frontend/noname/utils.rs [hunk 5]
```diff
@@ -0,0 +1,157 @@
...
+            parse_inputs(&ivc_inputs).map_err(|e| Error::Other(format!("{:?}", e)))?,
+            external_inputs,
+        ))
+    }
+}
+
+#[derive(Serialize, Deserialize)]
+pub struct NoNameExternalInputs {
+    pub external_inputs: Vec<usize>,
+}
```
Comment (arnaucube) [inline folding-schemes/src/frontend/noname/utils.rs:157]: I'm not finding any usage of this struct in the repo, is that correct? Maybe it can be removed (?)

File: examples/noname_full_flow.rs [hunk 6]
```diff
@@ -0,0 +1,154 @@
...
+
+    // set the external inputs to be used at each step of the IVC, it has length of 10 since this
+    // is the number of steps that we will do
+    let external_inputs = vec![
+        vec![Fr::from(8u32), Fr::from(2u32)],
+        vec![Fr::from(40), Fr::from(5)],
+    ];
+
+    // initialize the noname circuit
+    let f_circuit_params = (NONAME_CIRCUIT_EXTERNAL_INPUTS.to_owned(), 2, 2);
```
Comment (arnaucube) [inline examples/noname_full_flow.rs:52]: I was confused by the upper-case constant here thinking that was some constant of the logic, but after diving in I found out that this constant is in fact an example of a test circuit defined with Noname's language.
Since it only takes 4 lines, how would you see to paste here the circuit so the dev who reads the example sees at first sight that this is the circuit code?
Also in this way the test circuits of Noname don't have to be exposed as global constants in the frontend/noname package.

File: folding-schemes/src/frontend/noname/mod.rs (hunk 2 above)
Comment (arnaucube) [reply folding-schemes/src/frontend/noname/mod.rs:107]: Thanks for the explanation! This comment was useful ^^ how would you see adding it as an inline comment in the code for future readers (including ourselves)?

File: folding-schemes/src/frontend/noname/mod.rs [hunk 7]
```diff
@@ -0,0 +1,228 @@
...
+use ark_noname::utils::compile_source_code;
+use noname::backends::{r1cs::R1CS as R1CSNoName, BackendField};
+use noname::witness::CompiledCircuit;
+
+pub mod utils;
+
+use utils::*;
+
+#[derive(Debug, Clone)]
+pub struct NoNameFCircuit<F: PrimeField, BF: BackendField> {
```
Comment (arnaucube) [inline folding-schemes/src/frontend/noname/mod.rs:20]: is it `NoName` or `Noname`? from the repo readme it seems to be the latter (https://github.com/zksecurity/noname)

**Review by arnaucube — COMMENTED** [review]
Nice!! Very cool ^^

You can also add Noname under the "Available frontends" section in the README.md file now :smiley_cat:

File: folding-schemes/src/frontend/noname/utils.rs (hunk 4 above)
Comment (arnaucube) [reply folding-schemes/src/frontend/noname/utils.rs:23]: (this relates to this other comment: https://github.com/privacy-scaling-explorations/sonobe/pull/121#discussion_r1665803421 )

File: examples/noname_full_flow.rs [hunk 8]
```diff
@@ -0,0 +1,154 @@
...
+
+    // run n steps of the folding iteration
+    for (i, external_inputs_at_step) in external_inputs.iter().enumerate() {
+        let start = Instant::now();
+        nova.prove_step(rng, external_inputs_at_step.clone())
+            .unwrap();
+        println!("Nova::prove_step {}: {:?}", i, start.elapsed());
+    }
+
+    let rng = rand::rngs::OsRng;
```
Comment (CPerezz) [inline examples/noname_full_flow.rs:97]: We already have an `rng` defined above in line 77. This is not needed.

```suggestion
```

File: folding-schemes/Cargo.toml (hunk 1 above)
Comment (CPerezz) [reply folding-schemes/Cargo.toml:26]: How do we deal with this in the long run? PR to noname? An adaptor lib? Feels a bit weak to have this in a random repo

File: folding-schemes/Cargo.toml [hunk 9]
```diff
@@ -23,9 +23,13 @@ num-integer = "0.1"
 color-eyre = "=0.6.2"
 ark-bn254 = {version="0.4.0"}
 ark-groth16 = { version = "^0.4.0" }
+ark-noname = { git = "https://github.com/dmpierre/ark-noname", branch="feat/sonobe-integration" }
+noname = { git = "https://github.com/dmpierre/noname" }
+serde_json = "1.0.85"                                                                # to (de)serialize JSON
```
Comment (CPerezz) [inline folding-schemes/Cargo.toml:28]: It's weird to need the specific serde variant as a direct dep instead of serde itself only and the variant as dev-dep.

Can you double-check that is it needed?

File: folding-schemes/src/frontend/noname/mod.rs [hunk 10]
```diff
@@ -0,0 +1,228 @@
...
+        // compute witness using noname witness generator and outputs the next step
+        let external_inputs_json = match !external_inputs.is_empty() {
+            false => None,
+            true => {
+                let json =
+                    serde_json::to_string(&NoNameJSONExternalInputs::from(external_inputs.clone()))
+                        .map_err(|e| Error::JSONSerdeError(e.to_string()))?;
+                Some(json)
+            }
+        };
```
Comment (CPerezz) [inline folding-schemes/src/frontend/noname/mod.rs:66]: When you deal with binary option, you should use `if-else` instead of match.

File: folding-schemes/src/frontend/noname/mod.rs [hunk 11]
```diff
@@ -0,0 +1,228 @@
...
+            .generate_witness(ivc_inputs, noname_external_inputs)
+            .map_err(|e| Error::WitnessCalculationError(e.to_string()))?;
+
+        let mut assigned_z_i1 = vec![];
+        let z_i1_end_index = z_i.len() + 1;
+        for idx in 1..z_i1_end_index {
+            let value: BigUint = Into::into(noname_witness.witness[idx]);
+            let field_element = F::from(value);
+            assigned_z_i1.push(field_element);
+        }
```
Comment (CPerezz) [inline folding-schemes/src/frontend/noname/mod.rs:84]: Is Noname restricted to JSON?? If so, that's ok. Otherwise, I think we should use plain serde and allow the user to specify what does it want to use to deserialize stuff.

File: folding-schemes/src/frontend/noname/mod.rs (hunk 2 above)
Comment (CPerezz) [reply folding-schemes/src/frontend/noname/mod.rs:107]: I have a question here.

Why would we recieve FpVars in the first place? Can't we get plain field elements directly and we generate the vars if needed too? Seems less complex.

File: folding-schemes/src/frontend/noname/mod.rs [hunk 12]
```diff
@@ -0,0 +1,228 @@
...
+
+        Ok(assigned_z_i1)
+    }
+
+    fn generate_step_constraints(
```
Comment (CPerezz) [inline folding-schemes/src/frontend/noname/mod.rs:89]: As I always say in PRs. We need to make this fn as thin and fast as possible. So the less transformations and shenanigans we have on them the better.

Just saying to see if there's any way to simplify things.

File: folding-schemes/src/frontend/noname/mod.rs [hunk 13]
```diff
@@ -0,0 +1,228 @@
...
+            .map_err(|_| SynthesisError::Unsatisfiable)?;
+
+        let mut assigned_z_i1 = vec![];
+        let z_i1_end_index = z_i.len() + 1;
+        for idx in 1..z_i1_end_index {
+            // the assigned zi1 is of the same size than the initial zi and is located in the
+            // output of the witness vector
+            // we prefer to assign z_i1 here since (1) we have to return it, (2) we cant return
+            // anything with the `generate_constraints` method used below
+            let value: BigUint = Into::into(noname_witness.witness[idx]);
```
Comment (CPerezz) [inline folding-schemes/src/frontend/noname/mod.rs:126]: I assume we need the `BigUint` to be able to do from_string or similar no?
Have you checked if arkworks already has a way tio do so? Such that we don't need to deal with the BigUint?

Comment (CPerezz) [reply folding-schemes/src/frontend/noname/mod.rs:126]: Also, why the hell do they use String instead of bytes? At the end they are the same. But is much simpler to play with byte types anyway.

Do you happen to know?

File: folding-schemes/src/frontend/noname/utils.rs (hunk 4 above)
Comment (CPerezz) [inline folding-schemes/src/frontend/noname/utils.rs:23]: Why are these defined as string slices instead of using macros directly? Just wondering.

File: folding-schemes/src/frontend/noname/utils.rs [hunk 14]
```diff
@@ -0,0 +1,157 @@
...
+
+impl<F: PrimeField> From<Vec<F>> for NoNameJSONIVCInput {
+    fn from(value: Vec<F>) -> Self {
+        let mut ivc_input = Vec::with_capacity(value.len());
+        for field_element in value {
+            match field_element == F::zero() {
+                true => ivc_input.push("0".to_string()),
+                false => ivc_input.push(field_element.to_string()),
+            }
```
Comment (CPerezz) [inline folding-schemes/src/frontend/noname/utils.rs:42]: Why is this needed? Can't we convert `0` into `String`???

File: folding-schemes/src/frontend/noname/utils.rs [hunk 15]
```diff
@@ -0,0 +1,157 @@
...
+    }
+}
+
+impl NoNameJSONIVCInput {
+    pub fn from_fpvars<F: PrimeField>(value: &Vec<FpVar<F>>) -> Result<Self, SynthesisError> {
+        let mut field_elements = Vec::<F>::with_capacity(value.len());
+        for var in value {
+            let val = var.value()?;
+            field_elements.push(val);
+        }
```
Comment (CPerezz) [inline folding-schemes/src/frontend/noname/utils.rs:54]: I suggest to go for iterators rather than a for loop when the loops aren't crazy.

This makes it easier to add/remove and understand what happens there.

File: folding-schemes/src/frontend/noname/utils.rs [hunk 16]
```diff
@@ -0,0 +1,157 @@
...
+}
+
+impl<F: PrimeField> From<Vec<F>> for NoNameJSONExternalInputs {
+    fn from(value: Vec<F>) -> Self {
+        let mut external_inputs = Vec::with_capacity(value.len());
+        for field_element in value {
+            match field_element == F::zero() {
+                true => external_inputs.push("0".to_string()),
+                false => external_inputs.push(field_element.to_string()),
+            }
```
Comment (CPerezz) [inline folding-schemes/src/frontend/noname/utils.rs:66]: ditto

File: folding-schemes/src/frontend/noname/utils.rs [hunk 17]
```diff
@@ -0,0 +1,157 @@
...
+}
+
+impl NoNameJSONExternalInputs {
+    pub fn from_fpvars<F: PrimeField>(value: &Vec<FpVar<F>>) -> Result<Self, SynthesisError> {
+        let mut field_elements = Vec::<F>::with_capacity(value.len());
+        for var in value {
+            let val = var.value()?;
+            field_elements.push(val);
+        }
+        Ok(NoNameJSONExternalInputs::from(field_elements))
```
Comment (CPerezz) [inline folding-schemes/src/frontend/noname/utils.rs:79]: ditto

File: folding-schemes/src/frontend/noname/utils.rs [hunk 18]
```diff
@@ -0,0 +1,157 @@
...
+        let json_ivc_inputs = serde_json::from_str::<NoNameJSONIVCInput>(&ivc_public_inputs)
+            .map_err(|e| Error::JSONSerdeError(e.to_string()))?;
+
+        let json_external_inputs = match external_inputs {
+            Some(inputs) => Some(
+                serde_json::from_str::<NoNameJSONExternalInputs>(&inputs)
+                    .map_err(|e| Error::JSONSerdeError(e.to_string()))?,
+            ),
+            None => None,
+        };
```
Comment (CPerezz) [inline folding-schemes/src/frontend/noname/utils.rs:102]: Use `if let Some(X)` instead of this.

**Review by CPerezz — CHANGES_REQUESTED** [review]

File: folding-schemes/Cargo.toml (hunk 1 above)
Comment (dmpierre) [reply folding-schemes/Cargo.toml:26]: We already merged [a PR](https://github.com/zksecurity/noname/pull/137) in noname, which integrates this adaptor library into it directly. But there are two issues:
- Noname is using `arkworks=3.0` and an upgrade to `4.0` might not happen, [for now](https://github.com/zksecurity/noname/issues/138).
- We need to have an adapted version of this lib, accommodating the fact that the witness and public i/o is assigned *before* calling the lib.
This approach of having a fork of an adaptor lib is [similar to what we are doing with `ark-circom`](https://github.com/privacy-scaling-explorations/sonobe/blob/b5667968f48feacbccb1c3f5407f00f7e4892a4a/folding-schemes/Cargo.toml#L18). However, in this case, I hope to only fork `noname` in the future and get rid of `ark-noname`, as soon as `noname` moves on to `arkworks=4.0`.

File: folding-schemes/src/frontend/noname/mod.rs (hunk 2 above)
Comment (dmpierre) [reply folding-schemes/src/frontend/noname/mod.rs:107]: In sonobe, the `generate_step_constraints` method is used to compute the witness/trace of the `F` function, using the public state `z_i` and `external_inputs`. Both for design and security concerns (preventing users to allocate underconstrained witness values - cc @arnaucube, correct me if I'm wrong),`z_i` and `external_inputs` are allocated beforehand (when initializing the `AugmentedF` function). 
I agree that its kind of confusing (I asked about this to @arnaucube in the past!), but became convinced that's it's actually a nice way to do things! 
Still, if we wanted to get plain field elements, we will have to re-write the `FCircuit` trait and do a refactor of a few parts of the codebase.

File: folding-schemes/Cargo.toml (hunk 9 above)
Comment (dmpierre) [reply folding-schemes/Cargo.toml:28]: I am using `serde` to [derive the `Serialize, Deserialize` traits](https://github.com/privacy-scaling-explorations/sonobe/blob/d8e8f205296cca064edc943ce082d58aebe575ac/folding-schemes/src/frontend/noname/utils.rs#L11) and `serde_json` to [handle serializing and deserializing structs and json data](https://github.com/privacy-scaling-explorations/sonobe/blob/d8e8f205296cca064edc943ce082d58aebe575ac/folding-schemes/src/frontend/noname/mod.rs#L104). If I understand correctly your suggestions, wouldn't it be the other way around? Specifying `serde` as a dev dep and `serde_json` as a direct dependency?

File: folding-schemes/src/frontend/noname/utils.rs (hunk 14 above)
Comment (dmpierre) [reply folding-schemes/src/frontend/noname/utils.rs:28]: I know, that's weird, see [7f19f6b](https://github.com/privacy-scaling-explorations/sonobe/pull/121/commits/7f19f6b802b4dff3b40c47c170fb4c5089565d84)...

Comment (CPerezz) [reply folding-schemes/src/frontend/noname/utils.rs:28]: This is the problem of formatting as String instead of bytes I guess.. Can this be changed? Can we move away from the String type?

Comment (dmpierre) [reply folding-schemes/src/frontend/noname/utils.rs:28]: Nope, since we will then need to format those into a json that can be ingested by the noname witness generator..

File: folding-schemes/src/frontend/noname/mod.rs (hunk 2 above)
Comment (CPerezz) [reply folding-schemes/src/frontend/noname/mod.rs]: Ahh that's fair
If a refactor is needed let's ignore that for now. We can revisit this in the future then.

File: folding-schemes/Cargo.toml (hunk 9 above)
Comment (CPerezz) [reply folding-schemes/Cargo.toml:28]: Normally is in the other way. But in this case JSON is the only accepted format. So sadly we need to import on this way. 

It's a bit weird IMO. As we could have bincode and yaml too but they force JSON for whatever reason..

File: folding-schemes/Cargo.toml (hunk 3 above)
Comment (dmpierre) [reply folding-schemes/Cargo.toml]: [Corrected](https://github.com/privacy-scaling-explorations/sonobe/blob/d2db27d434dd70c5078cbd1738a9e8e6e0577c4d/folding-schemes/Cargo.toml#L33)

File: folding-schemes/src/frontend/noname/utils.rs (hunk 4 above)
Comment (dmpierre) [reply folding-schemes/src/frontend/noname/utils.rs]: [Changed](https://github.com/privacy-scaling-explorations/sonobe/blob/d2db27d434dd70c5078cbd1738a9e8e6e0577c4d/folding-schemes/src/frontend/noname/mod.rs#L127) here!

File: folding-schemes/src/frontend/noname/utils.rs (hunk 5 above)
Comment (dmpierre) [reply folding-schemes/src/frontend/noname/utils.rs]: Correct, removed it

File: examples/noname_full_flow.rs (hunk 6 above)
Comment (dmpierre) [reply examples/noname_full_flow.rs:57]: Yup, did it [here](https://github.com/privacy-scaling-explorations/sonobe/blob/d2db27d434dd70c5078cbd1738a9e8e6e0577c4d/examples/noname_full_flow.rs#L38)

File: folding-schemes/src/frontend/noname/mod.rs (hunk 7 above)
Comment (dmpierre) [reply folding-schemes/src/frontend/noname/mod.rs]: Changed all occurrences as requested

File: folding-schemes/src/frontend/noname/mod.rs [hunk 19]
```diff
@@ -0,0 +1,201 @@
+use crate::Error;
+use ark_noname::sonobe::NonameSonobeCircuit;
```
Comment (arnaucube) [inline folding-schemes/src/frontend/noname/mod.rs:2]: Seems weird to have the logic of `NonameSonobeCircuit` in a different repo (https://github.com/dmpierre/ark-noname/tree/caf7585aac5cc92fc319d649b0b564a8d9899fcf/src/sonobe.rs#L16), I was expecting this logic to be in the current file (`folding-schemes/src/frontend/noname/mod.rs`), but I'm not sure if I'm missing something.
Which would be the reason to have some sonobe specific logic in a separated `ark-noname`?

File: folding-schemes/src/frontend/noname/mod.rs (hunk 11 above)
Comment (dmpierre) [reply folding-schemes/src/frontend/noname/mod.rs]: You are right! I did a major refactor for this PR based on your remark. I removed most of the transformations I was doing. Now, the `generate_step_constraints` method is much more [lightweight](https://github.com/privacy-scaling-explorations/sonobe/blob/d2db27d434dd70c5078cbd1738a9e8e6e0577c4d/folding-schemes/src/frontend/noname/mod.rs#L75) and the content of the `utils.rs` file is [minimal](https://github.com/privacy-scaling-explorations/sonobe/blob/feat/noname-frontend/folding-schemes/src/frontend/noname/utils.rs).

File: folding-schemes/src/frontend/noname/mod.rs (hunk 12 above)
Comment (dmpierre) [reply folding-schemes/src/frontend/noname/mod.rs:75]: Agree, see the comment I did above, regarding the refactor of this PR

File: folding-schemes/src/frontend/noname/utils.rs (hunk 14 above)
Comment (dmpierre) [reply folding-schemes/src/frontend/noname/utils.rs]: Nope, we can't move away from the string type, since `Noname` expects field values to be represented as strings when parsing witness inputs, [see](https://github.com/zksecurity/noname/blob/cf9af0b07c248561985fa13107d267655e3619de/src/inputs.rs#L80)

File: folding-schemes/src/frontend/noname/utils.rs (hunk 15 above)
Comment (dmpierre) [reply folding-schemes/src/frontend/noname/utils.rs]: Yep, thanks for the suggestions, did this in the latest changes.

File: folding-schemes/src/frontend/noname/mod.rs (hunk 10 above)
Comment (dmpierre) [reply folding-schemes/src/frontend/noname/mod.rs]: Indeed, thanks for suggesting. Made sure this was the case in the latest PR change.

File: folding-schemes/src/frontend/noname/mod.rs (hunk 19 above)
Comment (dmpierre) [reply folding-schemes/src/frontend/noname/mod.rs:2]: This is similar to what we do with [`CircomCircuit`](https://github.com/privacy-scaling-explorations/sonobe/blob/d2db27d434dd70c5078cbd1738a9e8e6e0577c4d/folding-schemes/src/frontend/circom/mod.rs#L4), a struct from the `ark-circom` crate. However, since I have the hand on this crate, I created a separate module, specifically for sonobe :)

File: folding-schemes/src/frontend/noname/utils.rs (hunk 4 above)
Comment (dmpierre) [reply folding-schemes/src/frontend/noname/utils.rs]: Since those two noname circuit are only used for tests, isn't it more readable this way? How would you see things? (note that they are located [here](https://github.com/privacy-scaling-explorations/sonobe/blob/d2db27d434dd70c5078cbd1738a9e8e6e0577c4d/folding-schemes/src/frontend/noname/mod.rs#L127) now)

File: README.md [hunk 20]
```diff
@@ -33,6 +33,7 @@ Available frontends to define the folded circuit:
 
 - [arkworks](https://github.com/arkworks-rs), arkworks contributors
 - [Circom](https://github.com/iden3/circom), iden3, 0Kims Association
+- [Noname](https://github.com/zksecurity/noname/tree/main), a high-level programming language to write zkapps for kimchi or SnarkJS.
```
Comment (arnaucube) [inline README.md:36]: proposal to follow the style of the previous links, putting the authors of the repo instead of a custom description
```suggestion
- [Noname](https://github.com/zksecurity/noname/tree/main), zkSecurity
```

File: folding-schemes/src/frontend/noname/mod.rs (hunk 19 above)
Comment (arnaucube) [reply folding-schemes/src/frontend/noname/mod.rs:2]: I'm unsure: ark-circom was an already existing repo, which we updated to allow to work with other curves than BN254, and to support allocating the inputs, but there is nothing specifically related to Sonobe. And ideally we will end up opening a PR to the original ark-circom repo integrating those changes (which don't include any Sonobe specific stuff).
Whereas in ark-noname seems that there are specific Sonobe structs, which is what was making me think that we're splitting the Sonobe-related logic across the two different repos (also making ark-noname repo not Sonobe agnostic). Is that this way? (I've might be missing stuff)

Comment (dmpierre) [reply folding-schemes/src/frontend/noname/mod.rs:2]: The way we allocate the inputs is the same both in `ark-circom` and in `ark-noname`. In `ark-circom` the struct to do this is `CircomCircuit`, in `ark-noname` the struct is `NonameSonobeCircuit`. 
I would argue that, in both cases, the way we allocate inputs is specific to sonobe: another folding lib could have a different strategy for how it chooses to allocate those. That is why [what was merged in the noname repo](https://github.com/zksecurity/noname/pull/137) only consists in the arkworks-to-noname logic: it does not include the "specific" way we allocate inputs when working with sonobe. 
I guess that will be the same if the work done on the `ark-circom` repo is ever merged back upstream: it will most likely only include the work done on making it curve agnostic, but not the work we did to make it work with sonobe.
If that is confusing, I can change the name of the struct in the `ark-noname` lib to something that does not mention sonobe at all?

File: solidity-verifiers/Cargo.toml [hunk 21]
```diff
@@ -18,6 +18,7 @@ revm = {version="3.5.0", default-features=false, features=["std"]}
 rust-crypto = "0.2"
 num-bigint = "0.4.3"
 folding-schemes = { path = "../folding-schemes/"} # without 'light-test' enabled
+noname = { git = "https://github.com/dmpierre/noname" }
```
Comment (arnaucube) [inline solidity-verifiers/Cargo.toml:21]: Since this is only used for the example, it should be moved under the `dev-dependencies`.

**Review by arnaucube — APPROVED** [review]
Nice work!! ^^

File: folding-schemes/src/frontend/noname/utils.rs (hunk 14 above)
Comment (CPerezz) [reply folding-schemes/src/frontend/noname/utils.rs]: That's sad... Thanks mate!

File: folding-schemes/src/frontend/noname/utils.rs (hunk 4 above)
Comment (CPerezz) [reply folding-schemes/src/frontend/noname/utils.rs]: Let's leave it as is then. Is indeed more readable as you say.

**Review by CPerezz — APPROVED** [review]
LGTM!!!

Nice work!
